prov/gni/src/gnix_shmem.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2017 Cray Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#include "gnix_shmem.h"
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+
+#include "rdma/fi_errno.h"
+#include "gnix_util.h"
+
+int _gnix_shmem_create(
+	const char *path,
+	uint32_t size,
+	int (*init_func)(const char *path, uint32_t size, void *region),
+	struct gnix_shared_memory *region)
+{
+	int rc, fd, saved_errno, count;
+	void *buffer;
+
+	if (!region || !path) {
+		GNIX_WARN(FI_LOG_FABRIC, "bad params, ret=-FINVAL");
+		return -FI_EINVAL;
+	}
+
+	fd = open(path, O_CREAT | O_EXCL | O_RDWR, 0755);
+	if (fd >= 0) {
+		buffer = malloc(size);
+		if (!buffer) {
+			GNIX_INFO(FI_LOG_FABRIC,
+				"failed to allocate memory for shared memory segment");
+			return -FI_ENOMEM;
+		}
+
+		if (init_func) {
+			rc = init_func(path, size, buffer);
+			if (rc != 0) {
+				free(buffer);
+				return rc;
+			}
+		} else
+			memset(buffer, 0, size);
+
+		GNIX_INFO(FI_LOG_FABRIC,
+			"initializing shared memory segment, path=%s\n",
+			path);
+
+		count = write(fd, buffer, size);
+		if (count != size) {
+			GNIX_WARN(FI_LOG_FABRIC,
+				"failed to write to shared memory segment, "
+				"expected=%d actual=%d\n",
+				size, count);
+			saved_errno = errno;
+			close(fd);
+			unlink(path);
+			free(buffer);
+			errno = saved_errno;
+			return -saved_errno;
+		}
+		free(buffer);
+		fchmod(fd,
+			S_IRUSR | S_IWUSR |
+			S_IRGRP | S_IWGRP |
+			S_IROTH | S_IWOTH);
+	} else {
+
+		/* Unexpected error? */
+		if (errno != EEXIST) {
+			saved_errno = errno;
```

Comment:
I've found in other contexts with this type of approach, that if you have multiple failed system calls that can modify errno, that preserving the errno for the original error aids in debugging the actual problem.

```c
sz = failing_function();
if (sz == -1)
    /* let's assume that errno was something like enodev... 
        now to do something to report the error, but let's also 
         assume that it will also fail... */

   WARN(error_message, errno); 
   /* now depending on how WARN is implemented, it could 
       copy the reference to the stack at the correct time, or 
       it could make another failing call which modifies errno 
       prior to the actual printf, or other function. Because we 
       don't use a function with GNIX_WARN, it can't be safely 
       assumed that the value of errno will be safely saved on 
       the stack by the time the printf is done. In our debug 
       code, we perform sprintfs to get the node id, as well 
       as other things. This is only done to ensure everything 
       is saved safely. */

       // assume warn failed, and changed errno to ENOSPC, it may print the wrong message.

    /* Long story short: Preserving the errno here saves us 
        from implementation specific details of GNIX_WARN, 
        or other calls we might use here */
```